### ISSUE TITLE
Workspace builds might not rebuild correctly when dependecies are updated

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#215]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/215)
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jan 26, 2018, 15:16

After https://gitlab.com/BuildStream/buildstream/merge_requests/126 was landed, we realized that incremental workspace builds might not do the right thing in certain scenarios. Consider the following scenario:

- Element B depends on Element A
- User opens a workspace for B
- User builds B, so its workspace now has all the `.o` files and the underlying build system (`make` for instance) will not rebuild it on subsequent builds
- Now, the dependency A is updated so BuildStream will rebuild A and schedule B for rebuild too the next time one tries to build B
- Since A doesn't has a workspace associated with it, all the updated files in A will have a timestamp from 1970 within the sanbbox
- BuildStream will try to rebuild B but since it already has all the `.o` files will have a much newer timestamp than the header files form A, build systems like `make` won't do a rebuild

So to summarize, if the dependencies of a workspaced element are updated, they might not get rebuilt properly since the timestamps of all non-workspaced files will be ancient.

---

If someone comes across this issue, the immediate short-term solution would be to reset the workspace when your dependencies are updated. But we will probably need a better solution for the medium to long term. Perhaps we need a manifest of some sort but I'm not sure at present.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Jan 26, 2018, 15:21

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jan 29, 2018, 15:02

The only general solution that I can think of right now is to compare the file checksums of the staged dependencies between the previous successful build of that workspace and the current build. Setting the timestamps of the modified files to the current time should result in a proper incremental build.

The overhead of comparing the file checksums should not be prohibitive if we can use Merkle trees for the comparison, as that will be cheap for the case of few changes. OSTree uses Merkle trees but the internal ArtifactCache abstraction doesn't currently expose this and the tar artifact cache can't support this fast path.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 29, 2018, 15:12

> The only general solution that I can think of right now is to compare the file checksums of the staged dependencies between the previous successful build of that workspace and the current build. Setting the timestamps of the modified files to the current time should result in a proper incremental build.

This sounds like some interesting encoding of *"What the last build for this workspace built against"* in the workspace related metadata. What is interesting about this approach is that we can re-trigger the workspace rebuild to happen regardless if the not-matching-dependency was older or newer.

I.e. one might change the dependencies to a different artifact (using `bst track` or such), causing a rebuild; then one might *roll the same dependency back* and expect a rebuild again; the proposed solution treats this nicely by just considering that any non-matching artifact from the previous build is *"Now new"*.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jan 29, 2018, 17:55

> This sounds like some interesting encoding of _"What the last build for this workspace built against"_ in the workspace related metadata. What is interesting about this approach is that we can re-trigger the workspace rebuild to happen regardless if the not-matching-dependency was older or newer.

Yes, this was the idea. It's somewhat comparable to `git checkout` not touching unmodified files.

As a side-note, we will likely require information about the last build also for future support of incremental builds outside workspaces (see #21). We should keep this non-workspace use case in mind when designing the metadata structure.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jan 31, 2018, 10:20

I've been looking into this more closely and hit an issue. We use hardlinks for staging and thus must not modify the mtime of the (shared) inode as the same file may be staged multiple times across concurrent builds (or even in a single build).

The simplest option may be to create copies instead of hardlinks for affected files. A more efficient option may require additions in the FUSE layer, similar to what we will likely need for #38.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 1, 2018, 15:25

I suggest adding the following per-workspace metadata to `workspaces.yaml`:
 * Strong cache key of the last successful build of the workspace
 * List of staged files that have been updated in failed build attempts after
   the last successful build (empty if the last build was successful)

The file list is the simplest solution I've been able to come up with to
avoid issues with incremental builds after failed build attempts (which
may have built some but not all object files).

BuildStream needs to update the timestamps of all staged files that have
been changed since the last successful build and also for the files in the
above list. As mentioned in the previous comment, we must not update the
timestamps of hardlinks, so all affected files have to be copied until we
have a more elaborate solution.

To safe-guard against BuildStream terminating unexpectedly, the file
list should be written before the actual build job is started.

For efficient comparison of directory trees new API is required for
`ArtifactCache`. I suggest adding a `diff()` method with two cache key
parameters returning a list of files that have been changed or added
between the two artifacts. A third parameter to limit the comparison to
a subdirectory (`files`) would likely make sense.

The `OSTreeCache` implementation should be able to use the OSTree Merkle
tree to compare just the checksums instead of actual file content. No
such optimization will be possible for the tar-based artifact cache. It
could be discussed whether incremental build should be supported at all
without OSTree.

If the artifact cache does not contain the artifact for the last
successful build (or one of its dependencies), incremental build cannot
be used.

Incremental build is also unavailable when a previously closed workspace
is reopened as the metadata will be missing.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Feb 22, 2018, 09:37

created branch [`215-workspace-builds-might-not-rebuild-correctly-when-dependecies-are-updated`](https://gitlab.com/BuildStream/buildstream/compare/master...215-workspace-builds-might-not-rebuild-correctly-when-dependecies-are-updated)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Mar 20, 2018, 13:29

mentioned in merge request !315

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Mar 27, 2018, 14:43

closed via merge request !315
